src/control/server/instance_storage_rpc.go

```diff
@@ -156,3 +161,177 @@ func (ei *EngineInstance) StorageFormatNVMe() (cResults proto.NvmeControllerResu
+	if err != nil {
+		return errors.Wrapf(err, "retrieve health stats for %q (state %q)", ctrlr, state)
+	}
+	ctrlr.HealthStats = health
+
+	return nil
+}
+
+// Scan SMD devices over dRPC and reconstruct NVMe scan response from results.
+func scanEngineBdevsOverDrpc(ctx context.Context, ei *EngineInstance, pbReq *ctlpb.ScanNvmeReq) (*ctlpb.ScanNvmeResp, error) {
+	scanSmdResp, err := scanSmd(ctx, ei, &ctlpb.SmdDevReq{})
+	if err != nil {
+		return nil, errors.Wrap(err, "scan smd")
+	}
+
+	// Re-link SMD devices inside NVMe controller structures and populate scan response.
+
+	pbResp := ctlpb.ScanNvmeResp{
+		State: new(ctlpb.ResponseState),
+	}
+	seenCtrlrs := make(map[string]*ctlpb.NvmeController)
+
+	for _, sd := range scanSmdResp.Devices {
+		if sd.Ctrlr == nil {
+			return nil, errors.Errorf("smd %q has no ctrlr ref", sd.Uuid)
+		}
+
+		addr := sd.Ctrlr.PciAddr
+
+		if _, exists := seenCtrlrs[addr]; !exists {
+			c := new(ctlpb.NvmeController)
+			*c = *sd.Ctrlr
+			c.SmdDevices = nil
+			c.HealthStats = nil
+			seenCtrlrs[addr] = c
+			pbResp.Ctrlrs = append(pbResp.Ctrlrs, c)
+		}
+
+		c := seenCtrlrs[addr]
+
+		// Populate SMD (nvme-meta) if requested.
+		if pbReq.Meta {
+			nsd := new(ctlpb.SmdDevice)
+			*nsd = *sd
+			nsd.Ctrlr = nil
+			c.SmdDevices = append(c.SmdDevices, nsd)
+		}
+
+		// Populate health if requested.
+		if pbReq.Health && c.HealthStats == nil {
+			if err := populateCtrlrHealth(ctx, ei, sd.Uuid, c); err != nil {
+				return nil, err
+			}
+		}
+
+		// TODO: handle Basic request parameter.
+	}
+	ei.log.Debugf("seen ctrlrs: %+v", seenCtrlrs)
+
+	return &pbResp, nil
+}
+
+// bdevScanEngine calls either in to the private engine storage provider to scan bdevs if engine process
+// is not started, otherwise dRPC is used to retrieve details from the online engine.
+func bdevScanEngine(ctx context.Context, engine Engine, pbReq *ctlpb.ScanNvmeReq) (*ctlpb.ScanNvmeResp, error) {
+	ei, ok := engine.(*EngineInstance)
+	if !ok {
+		return nil, errors.New("not EngineInstance")
+	}
+
+	if pbReq == nil {
+		return nil, errors.New("nil request")
+	}
+
+	isUp := ei.IsStarted()
+	upDn := "down"
+	if isUp {
+		upDn = "up"
+	}
+	ei.log.Debugf("scanning engine-%d bdev tiers while engine is %s", ei.Index(), upDn)
+
+	if isUp {
+		return scanEngineBdevsOverDrpc(ctx, ei, pbReq)
+	}
+
+	// TODO: should anything be passed from pbReq here e.g. Meta/Health specifiers?
+
+	// Retrieve engine cfg bdevs to restrict scan scope.
+	req := storage.BdevScanRequest{
+		DeviceList: ei.runner.GetConfig().Storage.GetBdevs(),
+	}
+	if req.DeviceList.Len() == 0 {
+		return nil, errors.Errorf("empty device list for engine instance %d", ei.Index())
+	}
+
+	return bdevScanToProtoResp(ei.storage.ScanBdevs, req)
+}
+
+func smdQueryEngine(ctx context.Context, engine Engine, pbReq *ctlpb.SmdQueryReq) (*ctlpb.SmdQueryResp_RankResp, error) {
+	ei, ok := engine.(*EngineInstance)
+	if !ok {
+		return nil, errors.New("not EngineInstance")
+	}
```

Comment:
Converting the interface type instantly to the concrete type feels a bit kludgy to me. Is there a way to update the Engine interface to offer the methods needed here?

Reply:
I have actively been trying to reduce the size of the interface by following the advice in the code comments in the Engine interface definition in harness.go. Ideally we want less methods in the engine interface and not more.